client.go

```diff
@@ -1301,7 +1323,11 @@ func (c *HostClient) acquireConn() (*clientConn, error) {
 		return cc, nil
 	}
 	if !createConn {
-		return nil, ErrNoFreeConns
+		if !c.PreferWaitConn {
+			return nil, ErrNoFreeConns
+		}
+		<-c.connNotify
```

Comment:
I don't really like this implementation. This way if a conn doesn't free up before the deadline we still have to wait for one to free up before we continue here. So we might be waiting much longer than our timeout.

Reply:
i will reconsider the implementation, there can be a better choose. since timer is inefficiency, i made this choose because it will not blocking api ,since we certainly use DoTimeout rather than Do. in our ad serving,5120 conns  is set but sometime it still get a lot of no free conns error at monment.i do think it's necessary to take a action.

Reply:
You could just `select` between `c.connNotify` and a timer that signals at `deadline` (using `AcquireTimer`).